### ISSUE TITLE
ArgumentError: invalid byte sequence in UTF-8 when parsing

### ISSUE THREAD

**Issue by walro** [YorickPeterse/oga#143]
Hello,

I run into the problem in described in the title when parsing certain, unlucky, combinations of text. To reproduce:

```ruby
irb(main):026:0> xml = "<name>&amp;#2013266165;&amp;#2013265920</name>"
=> "<name>&amp;#2013266165;&amp;#2013265920</name>"
irb(main):027:0> Oga.parse_xml(xml)
ArgumentError: invalid byte sequence in UTF-8
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/entities.rb:83:in `gsub'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/entities.rb:83:in `decode'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/entity_decoder.rb:14:in `decode'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/entity_decoder.rb:5:in `try_decode'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/text.rb:24:in `text'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/character_node.rb:25:in `inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/node_set.rb:283:in `map'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/node_set.rb:283:in `inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/element.rb:245:in `block in inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/element.rb:238:in `each'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/element.rb:238:in `inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/node_set.rb:283:in `map'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/node_set.rb:283:in `inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/document.rb:92:in `block in inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/document.rb:88:in `each'
	from /Users/robin/.gem/ruby/2.2.3/gems/oga-1.3.1/lib/oga/xml/document.rb:88:in `inspect'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/cli/console.rb:14:in `run'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/cli.rb:308:in `console'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/cli.rb:10:in `start'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/bin/bundle:20:in `block in <top (required)>'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/lib/bundler/friendly_errors.rb:7:in `with_friendly_errors'
	from /Users/robin/.gem/ruby/2.2.3/gems/bundler-1.10.6/bin/bundle:18:in `<top (required)>'
	from /Users/robin/.gem/ruby/2.2.3/bin/bundle:23:in `load'
	from /Users/robin/.gem/ruby/2.2.3/bin/bundle:23:in `<main>'irb(main):028:0>
```

Line that raises: https://github.com/YorickPeterse/oga/blob/a938f23a0e5817b5924eff907d804ffaa23cfb8f/lib/oga/xml/entities.rb#L83

I understand that the text is bad and this might not really be fixable, but I think it would be great if a "better" error than ArgumentError could be raised.

Thanks!

**Comment by YorickPeterse** [comment]
@walro What kind of behaviour would you expect instead? I'm not a fan of catching ArgumentError and simply re-raising EncodingError as this doesn't help the user all that much.

**Comment by walro** [comment]
Thanks for getting back to me.

> @walro What kind of behaviour would you expect instead?

Ideally, no error at all :-). If decode fails for a single character it could return the character as-is (undecoded).

I started to investigate a bit, it seems like https://github.com/YorickPeterse/oga/blob/76b183e7abc78e42e4651cd3e37ced3e57b74cf3/lib/oga/xml/entities.rb#L77-L79 will convert `&#2013266165;` into an invalid character which later on causes the UTF-8 error. Maybe it's possible to test the validity before proceeding?

> I'm not a fan of catching ArgumentError and simply re-raising EncodingError as this doesn't help the user all that much

I think it could be a bit helpful if an error pertaining to this library would be raised (say `Oga::EncodingError`) rather than the more generic ArgumentError. For instance, the end user could introduce a somewhat "safer", less broad, rescue.

Interestingly enough I happened to notice that the reproduction steps does not trigger an error when using `pry` instead of `irb`, I can't figure out why. We have encountered the error in the wild though.

**Comment by YorickPeterse** [comment]
> Ideally, no error at all :-). If decode fails for a single character it could return the character as-is (undecoded).

Leaving the input unchanged upon a decoding failure seems to be the most logical solution, better than re-raising any kind of dedicated error.

As an aside, the following reproduces the problem described:

    ruby -r oga -e 'Oga.parse_xml("<name>&amp;#2013266165;&amp;#2013265920</name>").inspect'

**Comment by strcmp** [comment]
in such cases if a document cannot be encoded to the encoding it claims to be i think re-encoding the string as ASCII-8BIT would be an ok fallback solution. in this case it appears there is no document to infer an encoding from, and ASCII-8BIT should be assumed in such cases.

**Comment by strcmp** [comment]
```ruby
def self.parse_xml(xml)
  xml = xml.dup.force_encoding(:BINARY)
  ..
end
```

**Comment by YorickPeterse** [comment]
The fix for this is available in version 2.1 which should be available on RubyGems as of now.

**Comment by walro** [comment]
Thanks, seems to be working great!
